README.md
# EH-mdServer - a minimalist web server

This is a very simple web server which somewhat emulates GitPages for local
development.

## Installation

Install globally with NPM:

```console
npm install -g ehmdserver
```

Or, clone the repository and link it:

```console
git clone https://github.com/tdesposito/EH-mdServer.git
cd EH-mdServer
npm install
npm link
```

## Usage

```console
$ cd /my/project/dir
$ ehmdserver --port 3030 --root ./docs --alias /src ./src
ehMDserver: running at http://localhost:3030
```

## Command Line Options

| Option | Operation |
| -- | -- |
| --port, -p  | port number to listen on (default: 8000) |
| --alias, -a | add a url alias (see below) |
| --root, -r  | set the document root (defaults to cwd()) |
| --style, -s | set the stylesheet url for Markdown files (see more below) |
| --help, -h | display help |
| --version | displays the server version |

## Document Root and URL Aliases

By default, we look for all files in the current directory (where you started
the server, not where the server is installed).

If you want to specify the document root, use `--root /some/absolute/dir` or
`--root ./some/relative/dir`.

If hrefs (or Markdown links) in your source files reference things not under the
document root (such as your project's source files), you can add one or more URL
aliases to map URL prefixes to alternate file locations.

A common project might look like:

```text
/projectdir
  +- /src
     |- index.js
     |- utility.js
     |- README.md
  +- /docs
     +- index.md
     +- demo.html
```

If you start the server as `ehmdserver --root ./docs` from the project
directory, references to `/demo.html` inside of `index.md` will resolve
correctly, since they share a document root. A reference to `../src/README.md`
would fail; the `../` collapses and the browser will look for the README.md file
as `/src/README.md`.

To solve this, use:

`ehmdserver --root ./docs --alias /src ./src`

## Styling Rendered Markdown

We add a [simple CSS stylesheet](/md.css) to our rendered Markdown files. You
can override this by providing a (relative or absolute) URL for your preferred
stylesheet on the command line, such as `ehmdserver --style /static/my.css`.

* Credit to [Chris Patuzzo](https://twitter.com/chrispatuzzo) for the CSS.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

* The format is based on [Keep a Changelog](https://keepachangelog.com/)
* This project adheres to [Semantic Versioning](https://semver.org/)
* This project uses [cvbump](https://github.com/tdesposito/ChangelogVersionTool) to maintain this changelog.

## [1.1.1](https://github.com/tdesposito/EH-mdServer/releases/tag/v1.1.0) - 2022-01-28

### Fixed

* Fixed interface - now binds to 0.0.0.0 ([8a55225](https://github.com/tdesposito/EH-mdServer/commit/8a55225d2c9f325a8c59a2a455a677b9c60b24e1) by Todd Esposito)
* fix typo in sendError() ([40b6082](https://github.com/tdesposito/EH-mdServer/commit/40b60821d9b15830791362daca3f24bcb158b5be) by Todd D. Esposito)

## [1.1.0](https://github.com/tdesposito/EH-mdServer/releases/tag/v1.1.0) - 2022-01-25

### Added

* Added cvbump changelog ([47bb56b](https://github.com/tdesposito/EH-mdServer/commit/47bb56b429b2220287f944ed279cd392ea1d2769) by Todd Esposito)

### Other Updates

* Updated docs and sample ([33a01dd](https://github.com/tdesposito/EH-mdServer/commit/33a01dd833914dca040c992ca7115e55a41b7e57) by Todd Esposito)
* Updated default CSS to more complete version ([de01e31](https://github.com/tdesposito/EH-mdServer/commit/de01e3147ea3a3aac9f4f2a14a8e3093dda9bc5e) by Todd Esposito)
* update log output to be more standard (ibid) ([2c2d3ab](https://github.com/tdesposito/EH-mdServer/commit/2c2d3ab4107fe2f24e6f3e26bfe4f767ea30705a) by Todd Esposito)
* update log output to be more standard ([a71e3ae](https://github.com/tdesposito/EH-mdServer/commit/a71e3ae7d3caef592c2a9a486c11bafc239238ba) by Todd Esposito)
* update log output to be more standard ([bd07047](https://github.com/tdesposito/EH-mdServer/commit/bd070472b23f8e21907bfa23991a61d25bf8bab6) by Todd Esposito)

## [1.0.1](https://github.com/tdesposito/EH-mdServer/releases/tag/v1.0.1) - 2021-10-02

### Added

* add everything ([f524d47](https://github.com/tdesposito/EH-mdServer/commit/f524d4722ac359e29cef2f9e85e3b000a0213956) by Todd Esposito)

### Removed

* remove package cruft ([2c9b610](https://github.com/tdesposito/EH-mdServer/commit/2c9b610af2770b4400817fbb879944bef3e0b29c) by Todd Esposito)

## Notes

None at this time.


blog-version.js
const fs= require('fs')
const http = require('http')
const pathlib = require('path')
const showdown  = require('showdown')

showdown.setFlavor('github')


function resolvePath(req, rsp, path) {
  path = pathlib.normalize(path)
  if (fs.existsSync(path)) {
    if (fs.lstatSync(path).isFile())  {
      return sendFile(req, rsp, path)
    } else if (fs.lstatSync(path).isDirectory()) {
      return resolvePath(req, rsp, `${path.replace(/\/$/, '')}/index.html`)
    }
  } else if (path.endsWith('.html')) {
    // Maybe there's a corresponding .md file we can send instead?
    return resolvePath(req, rsp, path.slice(0, -5) + ".md")
  } else {
    return sendError(req, rsp, 404)
  }
}


function resolveUrl(url) {
  return pathlib.resolve(`./${url}`)
}


function sendError(req, rsp, code) {
  messages = {
    404: "Not Found"
  }
  rsp.writeHead(code, {'Content-Type': 'text/html'})
  rsp.end(`<html><body><h1>${code} ${messages[code]}</h1></body</html>`)
}


function sendFile(req, rsp, path) {
  if (path.endsWith('.md')) {
    return sendMarkdown(req, rsp, path)
  }

  body = fs.readFileSync(path)
  rsp.writeHead(200,
    {
      'Content-Type': 'text/plain',
      'Content-Length': Buffer.byteLength(body)
    }
  )
  rsp.end(body)
}


function sendMarkdown(req, rsp, path) {
  body = fs.readFileSync(path, 'utf-8')
  converter = new showdown.Converter()
  html = '<html><body>'
    + converter.makeHtml(body)
    + "</body></html>";
  rsp.writeHead(200,
    {
      'Content-Type': 'text/html',
      'Content-Length': Buffer.byteLength(html)
    }
  )
  rsp.end(html)
}


// create a simple HTTP server
const server = http.createServer((req, rsp) => {
  // convert the requested URL to a local pathname
  let pathname = resolveUrl(req.url)

  // send the content from that pathname, or an error
  return resolvePath(req, rsp, pathname)
})

// listen for requests.
server.listen(3000, "localhost", () => {
   console.log(`EH-mdServer: running at http://localhost:3000/`)
})


ehmdserver.js
#!/usr/bin/env node

// This file is part of EH-mdServer, Copyright (C) Todd D. Esposito 2021.
// Distributed under the MIT License (see https://opensource.org/licenses/MIT).

const fs = require('fs')
const http = require('http')
const pathlib = require('path')
const process = require('process')
const showdown  = require('showdown')
const yargs = require('yargs')

showdown.setFlavor('github')

const argv = yargs
  .option('port',
    {
      alias: 'p',
      description: 'port number to listen on',
      type: 'number',
      default: 8000,
    }
  ).option('alias',
    {
      alias: 'a',
      description: 'add a url alias (i.e. /static -> ../assets)',
      type: 'string',
      nargs: 2
    }
  ).option('root',
    {
      alias: 'r',
      description: 'set the document root (defaults to cwd())',
      type: 'string',
      default: process.cwd(),
    }
  ).option('style',
    {
      alias: 's',
      description: 'set the stylesheet URL for Markdown files',
      type: 'string',
      default: '/_ehmdserver/md.css',
    }
  ).help().alias('help', 'h').argv

// build a map for document locations
// start with aliases from the command line, if any
var docLocations = []
if (argv.alias) {
  for (i = 0; i < argv.alias.length; i += 2) {
    docLocations.push({
      'regex': new RegExp('^' + argv.alias[i]),
      'path': pathlib.resolve('./' + argv.alias[i + 1]),
    })
  }
}
// add the location of our internal resources
docLocations.push({
  'regex': new RegExp('^/_ehmdserver'),
  'path': __dirname,
})
// add the "docroot"
docLocations.push({
  'regex': new RegExp('^'),
  'path': argv.root,
})


function logRequest(req, rsp) {
  console.log(`${req.socket.remoteAddress} - - "${req.method} ${req.url}" ${rsp.statusCode} -`)
}


function resolvePath(req, rsp, path) {
  path = pathlib.normalize(path)
  if (fs.existsSync(path)) {
    if (fs.lstatSync(path).isFile())  {
      return sendFile(req, rsp, path)
    } else if (fs.lstatSync(path).isDirectory()) {
      return resolvePath(req, rsp, `${path.replace(/\/$/, '')}/index.html`)
    }
  } else if (path.endsWith('.html')) {
    // Maybe there's a corresponding .md file we can send instead?
    return resolvePath(req, rsp, path.slice(0, -5) + ".md")
  } else {
    return sendError(req, rsp, 404, "Not Found", `I couldn't resolve ${req.url}`)
  }
}


function resolveUrl(url) {
  for (i = 0; i < docLocations.length; i += 1) {
    if (docLocations[i].regex.test(url)) {
      return url.replace(docLocations[i].regex, docLocations[i].path)
    }
  }
}


function sendError(req, rsp, code, title, message) {
  rsp.writeHead(code, {'Content-Type': 'text/html'})
  rsp.end(`<html><body><h1>${code} ${title}</h1><p>${message}</p></body></html>`)
  logRequest(req, rsp)
  return true
}


function sendFile(req, rsp, path) {
  let ctype
  ext = path.split('/').slice(-1)[0].split('.').slice(-1)[0]
  switch (ext) {
    case 'css':
    case 'html':
      ctype = `text/${ext}`; break
    case 'gif':
    case 'jpeg':
    case 'png':
      ctype = `image/${ext}`; break
    case 'jpg':
      ctype = 'image/jpeg'; break
    case 'js':
      ctype = 'text/javascript'; break
    case 'md':
      return sendMarkdown(req, rsp, path)
    default:
      ctype = 'text/plain'
  }
  body = fs.readFileSync(path)
  rsp.writeHead(200,
    {
      'Content-Type': ctype,
      'Content-Length': Buffer.byteLength(body)
    }
  )
  rsp.end(body)
  logRequest(req, rsp)

  return true
}


function sendMarkdown(req, rsp, path) {
  body = fs.readFileSync(path, 'utf-8')
  converter = new showdown.Converter()
  html = '<html><head>'
    // + `<title>ehMDServer Presents...</title>`
    + `<link rel="stylesheet" href="${argv.style}"></style>`
    + '</head><body>'
    + converter.makeHtml(body)
    + "</body></html>";
  rsp.writeHead(200,
    {
      'Content-Type': 'text/html',
      'Content-Length': Buffer.byteLength(html)
    }
  )
  rsp.end(html)
  logRequest(req, rsp)

  return true
}


const server = http.createServer((req, rsp) => {
  if (req.method !== "GET") {
    return sendError(
      req, rsp, 405, "Method Not Allowed",
      "Microserver only responds to GET requests."
    )
  }
  return resolvePath(req, rsp, resolveUrl(req.url))
})

server.listen(argv.port, "0.0.0.0", () => {
   console.log(`ehMDserver: running at http://localhost:${argv.port}/`)
})
